src/message.cr

```diff
@@ -18,15 +17,88 @@ module Crylog
     getter datetime : Time
 
     # Any extra metadata added by a `Crylog::Processors::LogProcessor`.
-    property extra : Hash(String, Crylog::Context)
+    setter extra : Hash(String, Crylog::Context)?
 
-    # Any extra metadata added when the message was logged.
-    getter context : Hash(String, Crylog::Context)
+    @context : Hash(String, Crylog::Context)?
 
     # Represents the formatted version of this message.
     @[JSON::Field(ignore: true)]
     property formatted : String = ""
 
-    def initialize(@message : String, @context : Hash(String, Crylog::Context), @severity : Crylog::Severity, @channel : String, @datetime : Time, @extra : Hash(String, Crylog::Context)); end
+    # Lambda held for lazy message evaluation.
+    @[JSON::Field(ignore: true)]
+    @block : -> Logger::MsgType
+
+    @[JSON::Field(ignore: true)]
+    @set_message_context = false
+
+    def initialize(@severity : Crylog::Severity, @channel : String, &block : -> Logger::MsgType)
+      @datetime = Time.utc
+      @block = block
+    end
+
+    # The message that was logged.
+    def message : String
+      set_message_context
+      @message || ""
+    end
+
+    # Any extra metadata added when the message was logged.
+    def context : Hash(String, Crylog::Context)
+      set_message_context
+      @context.not_nil!
+    end
+
+    def extra
+      @extra ||= Hash(String, Crylog::Context).new
+    end
+
+    def message?
+      set_message_context
+
+      if msg = @message
+        !msg.empty?
+      else
+        false
+      end
+    end
+
+    def context?
+      set_message_context
+
+      if ctx = @context
+        !ctx.empty?
+      else
+        false
+      end
+    end
+
+    def extra?
+      if ext = @extra
+        !ext.empty?
+      else
+        false
+      end
+    end
+
+    # Resolves the passed block setting @message and @context.
+    def set_message_context : Nil
```

Comment:
A weird one off case with JSON formatter or potentially other macro generated formatters requires it's public use.  `message.to_pretty_json : message.to_json` won't see the variables if they aren't set and `.message` is never called so it's never set automatically.

If you know of a way to hook in to any macro's attempting direct instance var access I could use that.

Reply:
Could you not just call it in `#initialize`?

Reply:
I could make a public method `preserialize` which runs everything necessary for the serializers. `preserialize` would be documented and a public part of the API making `set_message_context` a private implementation detail.

Reply:
You mean `Message#initialize`?  No.  The block is only resolved if a handler handles the message.  Otherwise it's never called.  It's the main reason not logging is so much faster is the benchmarks above.  

You can compare a steady 13M cps regardless of arguments vs [8M, 4M, 2M, 5M] depending on arguments.  Anyone who uses `logger.debug { "User #{user.username} #{command} from #{request.remote_host}:#{request.remote_addr}" }` will see much greater improvements.  

One of my ported appications defering that block evaluation turned a 17s program in to a 0.4s program.  Ruby with deferred evaluation was ~6s.

Reply:
Ah fair enough.  Let's just do this instead:

```cr
# :nodoc:
def to_json(builder : JSON::Builder)
  set_message_context
  super
end
```

Then the end user wouldn't have to worry about calling that method and we can make it an implementation detail.